parlai/agents/example_seq2seq/example_seq2seq.py

```diff
@@ -150,107 +145,82 @@ def share(self):
 
         return shared
 
-    def predict(self, xs, ys=None, is_training=False):
-        """Produce a prediction from our model.
-
-        Update the model using the targets if available.
-        """
+    def v2t(self, vector):
+        """Converts vector to text."""
+        if vector.dim() == 1:
+            output_tokens = []
+            # Remove the final END_TOKEN that is appended to predictions
+            for i, token in enumerate(vector):
+                if token == self.END_IDX:
+                    break
+                else:
+                    output_tokens.append(token)
+            return [self.dict.vec2txt(output_tokens)]
+        elif vector.dim() == 2:
+            return [self.v2t(vector[i])[0] for i in range(vector.size(0))]
+        raise RuntimeError('Improper input to v2t with dimensions {}'.format(vector.size()))
+
+    def train_step(self, xs, ys, *args, **kwargs):
+        bsz = xs.size(0)
+        starts = self.START.expand(bsz, 1)  # expand to batch size
+        loss = 0
+        self.zero_grad()
+        self.encoder.train()
+        self.decoder.train()
+        target_length = ys.size(1)
+        # save largest seen label for later
+        self.longest_label = max(target_length, self.longest_label)
+
+        _encoder_output, encoder_hidden = self.encoder(xs)
+
+        # Teacher forcing: Feed the target as the next input
+        y_in = ys.narrow(1, 0, ys.size(1) - 1)
+        decoder_input = torch.cat([starts, y_in], 1)
+        decoder_output, decoder_hidden = self.decoder(decoder_input,
+                                                      encoder_hidden)
+
+        scores = decoder_output.view(-1, decoder_output.size(-1))
+        loss = self.criterion(scores, ys.view(-1))
+        loss.backward()
+        self.update_params()
+
+        _max_score, predictions = decoder_output.max(2)
+        return self.v2t(predictions)
+
+    def eval_step(self, xs, *args, **kwargs):
```

Comment:
The eval also needs `ys` in order to evaluate PPL? Maybe this method should be renamed `generate` and then `eval_step` can call `generate`.